doc/mgmt/gnmi/Config_Replace_gNOI.md

```diff
@@ -0,0 +1,226 @@
+|USB			   | Universal Serial Bus								|
+
+
+### 1. Overview 
+
+The Configuration Replacement feature provides the capability to replace the current running configuration, i.e., CONFIG_DB, with a known good configuration file saved in the serialization of DB format, i.e. config_db.json. 
+This feature is designed to limit the impact to features/services that are being replaced, with minimum service disruption. 
+
+### 2. Requirements
+#### Phase 1: 
+- Add YANG constraints and custom validation callbacks in SONIC yang specific to the features
+- Config replacement can be requested via gNOI management interfaces. 
+- Fail-safe operation - if an error is encountered, an error is reported and it should NOT take any partial config. i.e. transaction aborted without changing CONFIG_DB.
+- A syslog message will be generated in case of an error, including the table and attribute information causing the error.
+- Copying files to and from local directories and remote locations
+
+  | **Source**                 | **Destination**                        				|
+  |--------------------------|--------------------------------------------------------------------|
+  |config:<br>coredump:<br>ftp:<br>home:<br>http:<br>log:<br>running-configuration<br>scp:<br>startup-configuration<br>tech-support:<br>usb:<br>	   |config:<br>ftp:<br>home:<br>http:<br>running-configuration<br>scp:<br>startup-configuration<br>usb:<br>	|
+- Sorting, validation support for shallow and deep difference between new and current configuration.
+- Config replace operation should be performed with minimum disruption to the device, i.e. no container restart or system reboots expected.
+- Default configuration, e.g. copp_config.json, not to be saved to a serialized DB file should be preserved in the CONFIG_DB across config replacement operation.
+- Media format of intended config files can be a raw json, tar compresses
+- A new gNOI operation to perform config replacement.
+- CVL enforces YANG model constraints by libyang
+  
+#### Phase 2: Focus on the infrastructure part as defined below
+- Support all sonic yangs imported from sonic-mgmt-yang and implement custom validation where required
+
+### 3. Architecture Design 
+There are no changes to the current SONiC Architecture.
+
+### 4. High-Level Design 
+A new config replacement component provided by sonic-mgmt-common will be a key implemenation to orchestrate config replace operation requested. host-services module is responsible for getting the JSON file for configuration replacement. It will use below file transfer protocols to get the file from remote location.
+* SCP
+* FTP
+* HTTP
+
+Otherwise, It will get the file from below local space
+* config (/etc/sonic)
+* home (home directory of respective user)
+* USB
+
+If the input file is in compressed format(.tar), It will decompress the file to get the JSON format file. JSON file will be kept under temporary directory.
```

Comment:
>If the input file is in compressed format(.tar) [](http://example.com/codeflow?start=0&length=47)

I believe we do not need to invent a new gRPC protocol for this design purpose. The official gNMI protocol has "replace" operation defined, and when request "replace" with a path of "root" will be exactly what you need.

ref: [https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#344-modes-of-update-union-replace-replace-and-update](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#344-modes-of-update-union-replace-replace-and-update)